Add tests for requesting to join project

rmunn opened the pull request:
Fixes #1014

This adds two tests: one that exercises the create-project page workflow, and one that exercises the workflow that goes through the project page's "Ask to Join" button visible to org members.

rmunn commented:
This also contains a fix for #1409.

hahn-kev commented on frontend/tests/email/maildev-mailbox.ts:
```diff
@@ -14,9 +14,10 @@ export class MaildevMailbox extends Mailbox {
   }
 
-  async fetchEmails(subject: EmailSubjects): Promise<Email[]> {
+  async fetchEmails(subject: EmailSubjects, extraText?: string): Promise<Email[]> {
```
rather than `extraText` which is not a very descriptive name. Just make subject `EmailSubjects | string` that change should bubble up all the way to the tests

hahn-kev commented on frontend/tests/email/e2e-mailbox.ts:
```diff
@@ -12,9 +12,10 @@ export class E2EMailbox extends Mailbox {
   }
 
-  async fetchEmails(subject: EmailSubjects): Promise<Email[]> {
+  async fetchEmails(subject: EmailSubjects, extraText?: string): Promise<Email[]> {
```
same here

rmunn replied:
Done in commit 003d20a4.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +155,86 @@
+    let emailPage = await tempUserInTestOrg.mailbox.openEmail(page, EmailSubjects.VerifyEmail);
+    let pagePromise = emailPage.page.context().waitForEvent('page');
+    await emailPage.clickVerifyEmail();
+    let newPage = await pagePromise;
```
can we wrap this verification code into a helper since it's copied to the other tests?

rmunn replied:
Done in commit 003d20a4.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +155,86 @@
+
+    // Log in as manager, approve join request.
+    await loginAs(page.request, 'manager');
+    const managerMailbox = new MaildevMailbox('manager@test.com', page.request);
```
this isn't going to work in our staging tests unfortunately, this might mean that we need to make another test user which is a manager in order to receive the email.

rmunn replied:
Attempted fix in commit a8d9a37c, but it's not working yet and I've run out of time today. Will continue tomorrow using a8d9a37c as a starting point.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +155,86 @@
+
+    // Get to Sena-3 project page via org page, then ask to join
+    const testOrgPage = await new OrgPage(page, 'Test Org', testOrgId).goto();
+    await testOrgPage.membersTab.click();
```
why click the members tab?

rmunn replied:
Leftover code from making sure I got locators right. Removed in commit 003d20a4.

hahn-kev replied:
that didn't actually make it into that commit

rmunn replied:
Whoops; I was looking at the wrong function when I thought it was gone. It's now *actually* gone in commit f02d12b7.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +155,86 @@
+
+    // Log in as manager, approve join request.
+    await loginAs(page.request, 'manager');
+    const managerMailbox = new MaildevMailbox('manager@test.com', page.request);
```
same as above, this won't work on staging

rmunn replied:
Same as above, a8d9a37c tries to fix this but it isn't working yet on my machine.

hahn-kev commented on frontend/tests/fixtures.ts:
```diff
@@ -112,6 +114,29 @@
+    const tempUserId = await registerUser(page, name, email, password);
+    await loginAs(page.request, 'admin');
+    await addUserToOrg(page.request, tempUserId, testEnv.testOrgId, OrgRole.User);
+    await loginAs(page.request, email, password);
```
why do we login as the user being created? I'd rather keep this the same as tempUser where you're not logged in as them.

rmunn replied:
Because the corresponding `tempUser` fixture ends up with a login cookie for the just-created user (thanks to the `registerUser` call just before `await use(tempUser)`), and I wanted this one to behave the same way. But since this fixture has `loginAs('admin')` in between, it would result in *different* test behavior than `tempUser`, and I wanted to keep the same ends-up-logged-in behavior that `tempUser` has.

But if we don't want to document that `tempUser` ends up logged in as that user, then sure, we can get rid of this login as well. Let me know what you think.

hahn-kev replied:
no that sounds good, I didn't realize that the normal temp user worked that way, but that makes sense now

hahn-kev commented on frontend/tests/fixtures.ts:
```diff
@@ -112,6 +114,29 @@
+    const mailbox = await mailboxFactory();
+    const email = mailbox.email;
+    const name = `Test: ${testInfo.title} - ${email.replaceAll('@', "(at)")}`;
+    console.log(name);
```
forgot a console log

rmunn replied:
Removed in commit 003d20a4.

hahn-kev commented on frontend/tests/utils/authHelpers.ts:
```diff
@@ -37,6 +38,23 @@
   return userId;
 }
 
+export async function addUserToOrg(api: APIRequestContext, userId: string, orgId: string, role: OrgRole): Promise<UUID> {
```
this function probably isn't returning a `UUID` as `executeGql` just returns the raw json from the request which might be something like `{data:{changeOrgMemberRole:{organization:{id: 'value"}}}}` but it would only be returning the org id which was just passed in so I don't think we even care about the result.

rmunn replied:
Changed to `Promise<unknown>` in commit e342c9626.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +155,86 @@
+    await dashboardPage.openProject('Sena 3', 'sena-3');
+});
+
+test.only('ask to join project via project page', async ({ page, tempUserInTestOrg }) => {
```
this shouldn't be `test.only`

rmunn replied:
I had already fixed that but apparently didn't save before running `git commit`. Thanks for catching that anyway.

Fixed in commit 7c39cf5e.

hahn-kev reviewed (changes requested):
left some feedback. 

I tried running the tests and `join via project page` failed because it couldn't click the project link on the org page. I ran it multiple times and it never passed on other ff or chrome.

Running the `join via create project` test failed with this error:
```
Error: Timed out 5000ms waiting for expect(locator).toBeVisible()

Locator: locator('#group-extra-projects')
Expected: visible
Received: <element(s) not found>
Call log:
  - expect.toBeVisible with timeout 5000ms
  - waiting for locator('#group-extra-projects')


    at C:\dev\LexBox\frontend\tests\emailWorkflow.test.ts:178:51
```

rmunn commented:
Forgot to mention, when I ran this I had to mark the Sena-3 project as NON-confidential in order for the "Ask to join" button to show up. So I might need to create a whole separate, non-confidential, test project in order for this to work properly. Either that, or change our seeding data to make Sena-3 non-confidential, which I bet would require adjusting the expectations of a few other tests. I'll check what would be involved in that; it does make sense, from a "realism" point of view for Sena-3 (the project that's distributed with FieldWorks as example data) to be considered non-confidential rather than "confidentiality unknown".

rmunn commented:
Commit a8d9a37c is an attempt to fix the test by using Elawa (non-confidential in seeding data) instead of Sena-3 (confidentiality unknown in seeding data) so that the newly-added org member will be able to see the "Ask to Join" button. I haven't gotten it working, but I ran out of time for today. I'll continue working on this tomorrow.

hahn-kev commented on frontend/tests/email/mailbox.ts:
```diff
@@ -9,13 +9,13 @@ export interface Email {
   constructor(readonly email: string) { }
 
-  abstract fetchEmails(subject: EmailSubjects): Promise<Email[]>;
+  abstract fetchEmails(subject: EmailSubjects | string, extraText?: string): Promise<Email[]>;
```
need to remove extra text here no since you removed it from the other files

rmunn replied:
Fixed in commit 86137084, which also fixes the test failures (which were actually directly related to my forgetting to change the API in this abstract class: `extraText` was no longer used in the calls, but because I didn't change this API the compiler wasn't giving me errors).

rmunn commented:
On my local machine, the tests are now passing in Firefox but failing on Chromium. But when I run them in Chromium alone, they pass. That would seem to indicate an issue with running them in parallel, but I've double-checked all the obvious issues (such as #1409) and don't see anything. I'm going to push this and see what it does on GHA; it's possible that the tests are fine and I have a setup issue on my dev machine.

rmunn commented:
... But it doesn't *always* fail on Chromium? I ran the full test suite this time, and got *one* out of four tests (2 each on Firefox and Chromium) passing: "ask to join project via project page" on Chromium passed, the other three failed.

I'm going to throw up my hands at this point, say "If it passes on GHA, I'll call it good," and move on to a different PR.

rmunn commented:
@hahn-kev - GHA tests were failing because the `$lib` alias wasn't available. I decided in commit 84430e3d to replace the `OrgRole.User` enum with a hardcoded string to work around this problem. Obviously the better solution is to make the `$lib` alias available to Playwright in GHA so that if we ever change the value of `OrgRole.User` then the tests will automatically update accordingly. But I didn't want to spend even more time figuring out why the GHA tests aren't being run via Vite (which would make the `$lib` alias available), so I'm punting on that one.

That aside, these tests are now passing and I believe I've addressed all review concerns so far. Want to re-review this, and decide whether the `$lib` alias is important enough to hold off on merging the PR while we address it?

hahn-kev commented on frontend/tests/utils/authHelpers.ts:
```diff
@@ -37,6 +39,49 @@ export async function registerUser(page: Page, name: string, email: string, pass
+  return pagePromise;
+}
+
+export async function addUserToOrg(api: APIRequestContext, userId: string, orgId: string, role: string): Promise<unknown> {
```
I don't love doing the role this way, but I'm fine with it as a workaround. But could we restrict the role to valid values? eg: `'MANAGER' | 'EDITOR'`?

rmunn replied:
Done in commit 531bd7c5.

hahn-kev commented on frontend/tests/utils/authHelpers.ts:
```diff
@@ -37,6 +39,49 @@ export async function registerUser(page: Page, name: string, email: string, pass
+  `);
+}
+
+export async function addUserToProject(api: APIRequestContext, userId: string, projectId: string, role: string): Promise<unknown> {
```
same for roles here

rmunn replied:
Done in commit 531bd7c5.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +153,107 @@ test('register via new-user invitation email', async ({ page, mailboxFactory })
+    await newProjectPage.fillForm({name: 'Elaw', code: 'xyz', purpose: 'Testing', organization: 'Test Org'});
+    await expect(newProjectPage.extraProjectsDiv).toBeVisible();
+    await expect(newProjectPage.askToJoinBtn).toBeDisabled();
+    await newProjectPage.extraProjectsDiv.locator('#extra-projects-elawa-dev-flex').check();
```
can we make this a helper function on `CreateProjectPage`? something like `relatedProjectItem(projectCode: string): Locator` just because we've got a bit of magic strings here with the current locator.

rmunn replied:
Done in commit 531bd7c5.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +153,107 @@ test('register via new-user invitation email', async ({ page, mailboxFactory })
+    await dashboardPage.emailVerificationAlert.assertPleaseVerify();
+    let emailPage = await tempUserInTestOrg.mailbox.openEmail(page, EmailSubjects.VerifyEmail);
+    let pagePromise = emailPage.page.context().waitForEvent('page');
+    await emailPage.clickVerifyEmail();
```
should be using `verifyTempUserEmail` helper here

rmunn replied:
Done in commit 531bd7c5.

hahn-kev commented on frontend/tests/emailWorkflow.test.ts:
```diff
@@ -152,3 +153,107 @@ test('register via new-user invitation email', async ({ page, mailboxFactory })
+    let dashboardPage = await new UserDashboardPage(page).goto();
+
+    // Must verify email before being made manager of a project
+    await dashboardPage.emailVerificationAlert.assertPleaseVerify();
```
just a thought, since we're not actually testing the verification we could skip this. And since email verification does not require login we could actually skip logging in, going to the dashboard and checking for verification. The same can be done for the user and in the other test as well

rmunn replied:
Done in commit 531bd7c5.

hahn-kev reviewed (changes requested):
looks good, tests are passing for me now. There's a few minor things to change and I left a suggestion about streamlining email verification, but feel free to ignore it.

rmunn commented:
@hahn-kev - Addressed all review comments, should be ready for approval and merge now.

hahn-kev reviewed (approved):
looks good to me! nice work Robin
